requests-for-comments/2023-03-01-Temporal-RFC.md

```diff
@@ -0,0 +1,239 @@
+
+************************Stateful vs Stateless************************
+
+The practical difference between these two is that a stateless plugin generally executes in a deterministic way. This is mainly because it depends only on its inputs to execute. A stateful plugin on the other hand does not execute so deterministically. This is because of the fact that it depends on something outside of the inputs to the function, like an API. As an example let’s think of a plugin that during execution calls out to an api to either mutate or augment the data that is in flight. This is stateful because depending on the time that this plugin runs the results may vary. If this is a plugin that augments data with the response of an API the data returned by the API may change request to request. The api also may fail or stall. In our case most of these APIs are external to us and may fail or degrade at any point. This makes error handling of stateful plugins very important.
+
+- Plugins
+    - Stateless
+        - Transformations
+    - Stateful
+        - Exports
+        - API Calls
+        - Reverse ETLs
+        - Augmenters
+- Webhooks (Stateful)
+
+So what’s the pain point currently?
+
+Our biggest pain point currently is with the stateful plugins. Regardless of whether they are scheduled or realtime because they depend on external resources that can degrade or fail they require sophisticated timeout and retry logic that we don’t have currently. Another pain point is a relative lack of visibility into the system when things fail. The final pain point is that we don’t have the ability to easily compose these plugins together. What we need is a workflow framework. We are halfway through building one.
+
+You might ask yourself, James, if we are halfway through building a workflow engine, then why don’t we just finish it!? That’s a great question. The real reason we should not complete it is that frameworks already exist to do just this and building another one is not our competitive advantage against our competitors. Building one is a distraction from us building out our actual business.
+
+What is a compounding factor here is that this isn’t just a plugin server issue but an issue that is felt across the product: Async migrations, billing, and PoE are all tasks that fit a workflow engine really well. There will be other tasks that we need to run in the future that will need a solid framework for executing workflows as well. This allows us to focus on these and not something that is outside of our expertise and product space.
+
+# Temporal
```

Comment:
The other big workflow orchestration tool is Apache Airflow. I feel that a section comparing both would fit nicely.